azurerm/internal/services/network/resource_arm_virtual_network.go

```diff
@@ -106,9 +106,14 @@ func resourceArmVirtualNetwork() *schema.Resource {
 						},
 
 						"address_prefix": {
-							Type:         schema.TypeString,
-							Required:     true,
-							ValidateFunc: validate.NoEmptyStrings,
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"address_prefixes": {
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
```

Comment:
i think it would be nice to mimic https://github.com/openshift/terraform-provider-azurerm/commit/c7297566c7c606d4966464c489fcc98fb46e2b1a#diff-6286933fb2937fda2c74bce2ed7219e9R55-R66

1) it deprecates the previous field.
2) it makes it so that only one of them will be set.

Reply:
I initially did it this way. The problem occurs when you run terraform plan, or terraform apply. Because both fields exist, it will error out. I suppose the deprecation warning can be put there, just not the restriction that it has to be one or the other. Your thoughts?

Reply:
> The problem occurs when you run terraform plan, or terraform apply. Because both fields exist, it will error out
so why would both fields need to exist together? only one of them should be supplied in the correct case.

Reply:
My lack of terraform knowledge is showing here. For reasons not known to me at the moment, when terraform is refreshing state, it always reads back both address_prefix and address_prefixes. I probably need to understand why that is happening before changing this to allow only one or the other. address_prefix comes back as null, and address_prefixes gets the proper values. I will investigate and update accordingly.

Reply:
I got to the bottom of the problem here. Since these values are a sub-schema, they needed to be prefixed by the top level schema. Fixed!

Reply:
https://github.com/hashicorp/terraform-plugin-sdk/issues/71

Reply:
@jhixson74  ^^

i don't think `subnet.0.<>` would actually work

Reply:
After digging into this further, subnet.0 is only referencing the first subnet block. There does not seem to even be a way to reference elements in a set like this. I've dug through a lot of terraform providers to find any that do this, and not a single one of them does. The few similar cases all just use 0 like here, which we know to be incorrect. The URL you link expresses the issue, but no resolution. In some of the test case code, where there are sets, '#' is used to check against the index number. That is all I have come up with so far. Now to try and find a way around this,  that is hopefully a good fix ;-)

Reply:
> The few similar cases all just use 0 like here, which we know to be incorrect.

It's not necessarily incorrect, it's just a stronger restriction. Rather than "each element can only have one of `address_prefix` or `address_prefixes`", it's requiring "the array as a whole can either have `address_prefix` in every element or `address_prefixes` in every element, but it can't mix and match".